src/bartiq/compilation/preprocessing.py

```diff
@@ -0,0 +1,139 @@
+from collections import defaultdict
+from dataclasses import replace
+from typing import Callable, TypeVar
+
+from .._routine import Constraint, PortDirection, Resource, ResourceType, Routine
+from ..compilation._utilities import is_single_parameter
+from ..symbolics.backend import SymbolicBackend, TExpr
+
+T = TypeVar("T")
+
+PreprocessingStage = Callable[[Routine[T], SymbolicBackend[T]], Routine[T]]
+
+
+def postorder_transform(transform: PreprocessingStage[T]) -> PreprocessingStage[T]:
+    def _inner(routine: Routine[T], backend: SymbolicBackend[T]) -> Routine[T]:
+        return transform(
+            replace(routine, children={child.name: _inner(child, backend) for child in routine.children.values()}),
+            backend,
+        )
+
+    return _inner
+
+
+@postorder_transform
+def add_default_additive_resources(routine: Routine[T], backend: SymbolicBackend[T]) -> Routine[T]:
+    child_resources_map: defaultdict[str, set[str]] = defaultdict(set)
+
+    for child in routine.children.values():
+        for resource in child.resources.values():
+            if resource.type == ResourceType.additive:
+                child_resources_map[resource.name].add(child.name)
+
+    additional_resources: dict[str, Resource[T]] = {
+        # The default here is to satisfy the typechecker
```

Comment:
I don't understand this comment.

Reply:
Well, that's because it should be there anymore 😄 Fixed in [a4d62fd](https://github.com/PsiQ/bartiq/pull/115/commits/a4d62fd9faa37f5d4c3ff401d04f2884227a16f2)